contribs/drt/src/main/java/org/matsim/contrib/drt/routing/DefaultDrtRouteUpdater.java

```diff
@@ -67,19 +72,28 @@ public DefaultDrtRouteUpdater(DrtConfigGroup drtCfg, Network network,
 
 	@Override
 	public void notifyReplanning(ReplanningEvent event) {
-		Stream<Leg> drtLegs = StreamEx.of(population.getPersons().values())
-				.flatMap(p -> p.getSelectedPlan().getPlanElements().stream())
-				.select(Leg.class)
-				.filterBy(Leg::getMode, drtCfg.getMode());
-		executorService.submitRunnablesAndWait(drtLegs.map(l -> (router -> updateDrtRoute(router, l))));
+		List<Future<?>> futures = new LinkedList<>();
+		
+		for (Person person : population.getPersons().values()) {
+			for (Trip trip : TripStructureUtils.getTrips(person.getSelectedPlan())) {
+				for (Leg leg : trip.getLegsOnly()) {
+					if (leg.getMode().equals(drtCfg.getMode())) {
+						futures.add(executorService.submitRunnable(router -> updateDrtRoute(router, person, trip.getTripAttributes(), leg)));
+					}
+				}
+			}
+		}
+		
+		futures.forEach(Futures::getUnchecked);
```

Comment:
I tried a bit around, but I couldn't come up with a more elegant solution to pass a list of arguments from all levels of the person -> trip -> leg hierarchy to the router. Maybe there is a better way?

Reply:
Looks nice and readable 👍